docs/ql-design-patterns.md

```diff
@@ -0,0 +1,82 @@
+# CodeQL Design Patterns
+
+A list of design patterns you are recommended to follow.
+
+## `::Range` for extensibility and refinement
+
+To allow both extensibility and refinement of classes, we use what is commonly referred to as the `::Range` pattern (since https://github.com/github/codeql/pull/727), but the actual implementation can use different names.
+
+<details>
+<summary>Generic example of how to define classes with ::Range</summary>
+
+Instead of
+```ql
+/** <QLDoc...> */
+abstract class MySpecialExpr extends Expr {
+  /** <QLDoc...> */
+  abstract int memberPredicate();
+}
+```
+with
+```ql
+class ConcreteSubclass extends MySpecialExpr { ... }
+```
+
+use
+
+```ql
+/**
+ * <QLDoc...>
+ *
+ * Extend this class to refine existing API models. If you want to model new APIs,
+ * extend `MySpecialExpr::Range` instead.
+ */
+class MySpecialExpr extends Expr {
+  MySpecialExpr::Range range;
+
+  MySpecialExpr() { this = range }
+
+  /** <QLDoc...> */
+  int memberPredicate() { result = range.memberPredicate() }
+}
+
+/** Provides a class for modeling new <...> APIs. */
+module MySpecialExpr {
+  /**
+   * <QLDoc...>
+   *
+   * Extend this class to model new APIs. If you want to refine existing API models,
+   * extend `MySpecialExpr` instead.
+   */
+  abstract class Range extends Expr {
+    /** <QLDoc...> */
+    abstract int memberPredicate();
+  }
+}
+```
+with
+```ql
+class ConcreteSubclass extends MySpecialExpr::Range { ... }
+```
+
+</details>
+
+Note, previously we used to write `MySpecialExpr() { this = self }`, but we now recommend using `MySpecialExpr() { this = range }` instead to avoid anyone mistakenly thinking that `self` and `this` are synonyms in general.
```

Comment:
```suggestion
Note that in some libraries, the `range` field is in fact called `self`. While the name of the field does not matter, using `range` may be less confusing in contexts like Python analysis, where `self` is by strong convention how you refer to the receiver of a method -- in QL, we denote that with the `this` keyword.
```

Reply:
I'm not 100% on board with your suggested change. I would like to put more emphasis on `range` being the recommended name for the field. How about this adapted version?

```suggestion
Note that in some libraries, the `range` field is in fact called `self`. While we do recommend using `range` for consistency, the name of the field does not matter (and using `range` avoids confusion in contexts like Python analysis that has strong usage of `self`).
```

Coming back to this, I'm realizing that the audience of this document is probably fairly invested in the codeql repo anyways, so I believe it's unlikely anyone will be confused about `self`/`this. I removed that part completely at some point, but added a small note about it again.

Reply:
@p0 you marked this as resolved, but didn't apply my suggestion. Your thumbs up makes me believe you also wanted the change I suggested?

Reply:
Thanks for spotting that -- I guess I was assuming you'd click 'commit' 👍